### ISSUE TITLE
Charactistic value guarding

### ISSUE THREAD

**Issue by snowdd1** [homebridge/HAP-NodeJS#141]
Quite a lot of characteristics have a restricted list of compliant values, such as 
-  [Characteristic.AirParticulateSize](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L56)
-  [Characteristic.AirQuality](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js#L75)
  and many more.

As I have seen a lot of user issues with "vanishing" HAP-nodeJS accessories and homebridges ( @nfarina ), that apparently were blocked by homekit after sending non-compliant values, I would start a (hopefully short!) discussion about how to guard `Characteristic.setValue()` against trying to set non compliant values.

Of course each accessory programmer could do it manually himself, but I would assume there would be an easier way.
For Chracteristics with value ranges we have the props object. For Characteristics with distinct values, we have the value lists as items of the charcteristic object. I am not deep enough into JavaScript to see a way to extract them programmatically and failsafe, as long as they do not have a common Name, such as
`Characteristic.values.AirQuality.UNKNOWN`
so that one could check for `(Characteristic.value)` existence and then all values of members of that object.

Any better ideas?

**Comment by justme-1968** [comment]
maybe both types can be guarded by checking against min and max. the enumeration types are all continuous as far as i can see. so setting max and min and checking against them should work.
one open point is if values outside of this range should be reported as errors ir silently adjusted.

i think it would also be good to enforce correct step values by something like this:
```
value = Math.round(value / step) * step;
```
as i had 'vanishing' accessories if for example if the target temperature was 22.3 instead of 22.5 for a step value is 0.5.

**Comment by cflurin** [comment]
This was also an issue with ```Characteristic.prototype.toHAP``` in context with ```characteristic.setProps```. Nike added a check in Characteristc.js:
```javascript
  // ensure our value fits within our constraints if present
  var value = this.value;
  if (this.props.minValue != null && value < this.props.minValue) value = this.props.minValue;
  if (this.props.maxValue != null && value > this.props.maxValue) value = this.props.maxValue;
```

**Comment by snowdd1** [comment]
AH, I didn't notice that it was limited now in [toHAP](https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/Characteristic.js#L211). So it doesn't get logged as out-of-bounds.
However, did anyone try to set min adn max for characteristics in which it was not planned by Apple? Are they still compliant, otherwise we would need "hidden" min and max values in properties?

**Comment by cflurin** [comment]
Funny, I quickly tried to change Target Horizontal Tilt Angle in the HKA Simulator:
The defaults are:  min: -90 - max: 90. If you change the min e.g. to -30 the Simulator says:
```
Min/Max must be between 0 and 360 for arc degrees
```

**Comment by snowdd1** [comment]
@justme-1968 
> maybe both types can be guarded by checking against min and max. the enumeration types are all continuous as far as i can see. so setting max and min and checking against them should work.
> one open point is if values outside of this range should be reported as errors ir silently adjusted.
> 
> i think it would also be good to enforce correct step values by something like this:
> 
> value = Math.round(value / step) * step;
> 
> 
> as i had 'vanishing' accessories if for example if the target temperature was 22.3 instead of 22.5 for a step value is 0.5.

I just had this with my complete homebridge with all room asignments gone... :cry: 
Although I had a guard in place:
`hk_value = Math.round(val/characteristic.props.minStep)/(1/characteristic.props.minStep);`
with a minStep value of 0.1...
And node made that: `val` was 22.36 °C 
````
> Math.round(22.36/0.1)*0.1
22.400000000000002
````
and everything gone.

Interesting however, that a 
````
> Math.round(22.36/0.1)/(1/0.1)
22.4
```` 
gives the proper value. Any information scientists here to explain this?

**Comment by snowdd1** [comment]
After having settled the minStep issue in my platform, I now went for the fixed value characteristics.
As my explanation of what to achieve was hard to understand, I have prepared a little snipped of code in one of my branches to test and to discuss.

https://github.com/snowdd1/homebridge/blob/Playground_Webserver/accessories/knxdevice.js#L84-L133

This little function `safeSet(char, value)` automatically searches for value properties of a Characteristic template with the same displayName  (weak connection I know, but I knew of none better) and stores them in an array within the current Characteristic `char`, if not existing already.

After that, it compares the `value` passed with the array and if it doesn't find the value defaults to the first allowed value from the array.

I tried this successfully with `PositionState` characteristic, throwing values 0-255 at it.

Would be much easier if the allowed values were easier to get from a characteristic instance. And then, this check could be made within `Characteristic.toHAP()` instead of every accessory.

Any more thoughts on that?
